src/noit_metric_director.c

```diff
@@ -280,47 +413,70 @@ noit_metric_director_flush(eventer_t e) {
 
 static void
 distribute_metric(noit_metric_message_t *message) {
-  void *vhash, *vinterests;
-  caql_cnt_t *interests = NULL;
-  char uuid_str[UUID_STR_LEN + 1];
-  mtev_uuid_unparse_lower(message->id.id, uuid_str);
-  if(mtev_hash_retrieve(&id_level, (const char *) &message->id.id, UUID_SIZE,
-      &vhash)) {
-    if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
-        message->id.name_len_with_tags, &vinterests)) {
-      interests = vinterests;
-      distribute_message_with_interests(interests, message);
+  int i;
+  static uuid_t uuid_zero = { 0 };
+  void *vhash, **vinterests;
+  interest_cnt_t has_interests = 0;
+  interest_cnt_t interests[nthreads];
+  memset(interests, 0, sizeof(interests));
+
+
+  mtev_memory_begin();
+  /* First we process interests for specific metrics: uuid-metric|ST[tags...] */
+  for(i=0; i<2; i++) {
+    /* First time through we use the message's ID, second time we use the NULL UUID
+     * which is used to "match any" check.
+     */
+    if(mtev_hash_retrieve(&id_level, i ? (const char *)&uuid_zero : (const char *)&message->id.id, UUID_SIZE,
+        &vhash)) {
+      if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
+                            message->id.name_len_with_tags, (void **)&vinterests)) {
+        memcpy(interests, *vinterests, sizeof(interests)); /*  we access this exactly once, no need to ck load or tmp variable */
+        for(i=0; i<nthreads; i++) has_interests |= interests[i];
+      }
+      if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
+                            message->id.name_len, (void **)&vinterests)) {
+        memcpy(interests, *vinterests, sizeof(interests)); /*  we access this exactly once, no need to ck load or tmp variable */
+        for(i=0; i<nthreads; i++) has_interests |= interests[i];
+      }
     }
   }
 
+  /* Next we run search queries per lane to find matches */
+  for(i=0; i<nthreads; i++) {
+    if(interests[i] != 0) continue; /* already interest, skip any work */
+    struct search2 *search;
+    ck_hs_iterator_t iter = CK_HS_ITERATOR_INITIALIZER;
+    while(ck_hs_next_spmc(&search_asts[i].hs, &iter, (void **)&search)) {
```

Comment:
wasn't a spin lock put in place to protect this hash table?

Reply:
yes... for modification